### PR TITLE
Implement pooling layer

### LINKED ISSUE #17: Implement pooling layer

### PR DESCRIPTION
**Pull request by NeiroYT**
Closes #17

### CODE REVIEW

File: include/layers/PoolingLayer.hpp [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+PoolingLayer<ValueType>::PoolingLayer(const Shape& input_shape,
+                                      const Shape& pooling_shape,
+                                      const std::string& pooling_type)
+    : poolingShape_(pooling_shape), Layer<ValueType>(input_shape, input_shape) {
+  if (pooling_shape.dims() > input_shape.dims()) {
+    throw std::invalid_argument("Pooling dims is bigger than the input dims");
+  }
+  if (pooling_shape.dims() > 2 || pooling_shape.dims() == 0) {
+    throw std::invalid_argument("Pooling dims is bigger than 2 or is 0");
+  }
```
Comment (aobolensk) [inline include/layers/PoolingLayer.hpp:50]: It is better to split this into 2 conditions

File: include/layers/PoolingLayer.hpp [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+  if (pooling_type == "average") {
+    poolingType_ = 0;
+  } else if (pooling_type == "max") {
+    poolingType_ = 1;
+  } else {
+    throw std::invalid_argument("No such pooling type");
```
Comment (aobolensk) [inline include/layers/PoolingLayer.hpp:56]:
```suggestion
    throw std::invalid_argument("Pooling type " + pooling_type + " is not supported");
```

File: src/layers/Layer.cpp [hunk 3]
```diff
@@ -0,0 +1,19 @@
+#include "layers/Layer.hpp"
+
+size_t Shape::get_index(const std::vector<size_t>& coords) const {
+  if (coords.size() != dims_.size()) {
+    throw std::invalid_argument("Invalid index vector");
+  }
+  size_t res = 0;
+  size_t mulbuf;
+  for (size_t i = 0; i < coords.size(); i++) {
+    // to get to the i line
```
Comment (aobolensk) [inline src/layers/Layer.cpp:10]:
```suggestion
    // to get to the ith line
```

File: src/layers/Layer.cpp [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+    mulbuf = std::accumulate(dims_.begin() + (i + 1), dims_.end(), size_t(1),
+                             std::multiplies<>());
+    if (coords[i] >= dims_[i]) {
+      throw std::out_of_range("Out of shape");
```
Comment (aobolensk) [inline src/layers/Layer.cpp:14]:
```suggestion
      throw std::out_of_range("Out of range");
```

File: include/layers/PoolingLayer.hpp [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+  if (pooling_shape.dims() > input_shape.dims()) {
+    throw std::invalid_argument("Pooling dims is bigger than the input dims");
+  }
+  if (pooling_shape.dims() > 2 || pooling_shape.dims() == 0) {
+    throw std::invalid_argument("Pooling dims is bigger than 2 or is 0");
+  }
+  if (pooling_type == "average") {
+    poolingType_ = 0;
+  } else if (pooling_type == "max") {
+    poolingType_ = 1;
```
Comment (aobolensk) [inline include/layers/PoolingLayer.hpp:54]: Introduce and use enum here for pooling type

File: include/layers/PoolingLayer.hpp [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+        }
+        switch (poolingType_) {
+          case 0:
+            res.push_back(avg_pooling(pooling_buf));
+            break;
+          case 1:
+            res.push_back(max_pooling(pooling_buf));
+            break;
+          default:
+            throw std::runtime_error("Unhandled exception");
```
Comment (aobolensk) [inline include/layers/PoolingLayer.hpp:107]: Did you mean "unknown pooling type"?

File: include/layers/PoolingLayer.hpp [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+  size_t tmpheight = 0;
+  size_t pshape1;
+  // O(N^2)
+  if (this->inputShape_.dims() != 1) {
+    if (poolingShape_.dims() == 1) {
+      pshape1 = 1;
+    } else {
+      pshape1 = poolingShape_[1];
+    }
+    for (size_t i = 0; i < this->outputShape_[0]; i++) {
```
Comment (aobolensk) [inline include/layers/PoolingLayer.hpp:83]: For implementation part it is better to extract checks to the function like "isOutOfBounds". The code will be more readable and checks will be simplified across the whole function

**Review by aobolensk — CHANGES_REQUESTED** [review]

File: include/layers/PoolingLayer.hpp (hunk 5 above)
Comment (NeiroYT) [reply include/layers/PoolingLayer.hpp:54]: Ok

File: include/layers/PoolingLayer.hpp (hunk 7 above)
Comment (NeiroYT) [reply include/layers/PoolingLayer.hpp:83]: I'll try to fix readability here

File: include/layers/PoolingLayer.hpp (hunk 6 above)
Comment (NeiroYT) [reply include/layers/PoolingLayer.hpp:107]: Fixed

File: src/layers/Layer.cpp [hunk 8]
```diff
@@ -0,0 +1,19 @@
+#include "layers/Layer.hpp"
+
+size_t Shape::get_index(const std::vector<size_t>& coords) const {
```
Comment (allnes) [inline src/layers/Layer.cpp:3]: if you refactor FCLayer, please move `mat_vec_mul` function to FCLayer too

Comment (NeiroYT) [reply src/layers/Layer.cpp:3]: Ok

**Review by aobolensk — APPROVED** [review]

**Review by allnes — APPROVED** [review]

**Review by Semyon1104 — APPROVED** [review]

**Pull request merged**